feat(satp-hermes): update satp to 0.0.3-beta

RafaelAPB opened the pull request:
**Pull Request Requirements**
- [x] Rebased onto `upstream/main` branch and squashed into single commit to help maintainers review it more efficient and to avoid spaghetti git commit graphs that obfuscate which commit did exactly what change, when and, why.
- [x] Have git sign off at the end of commit message to avoid being marked red. You can add `-s` flag when using `git commit` command. You may refer to this [link](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) for more information.
- [x] Follow the Commit Linting specification. You may refer to this [link](https://www.conventionalcommits.org/en/v1.0.0-beta.4/#specification) for more information. 

**Character Limit**
- [x] Pull Request Title and Commit Subject must not exceed 72 characters (including spaces and special characters).
- [x] Commit Message per line must not exceed 80 characters (including spaces and special characters).

**A Must Read for Beginners**
For rebasing and squashing, here's a [must read guide](https://github.com/servo/servo/wiki/Beginner's-guide-to-rebasing-and-squashing) for beginners.

RafaelAPB commented on packages/cactus-plugin-satp-hermes/src/main/typescript/cross-chain-mechanisms/oracle/oracle-scheduler-manager.ts:
```diff
@@ -0,0 +1,158 @@
+      throw new Error(`Poller with id "${id}" already exists.`);
+    }
+
+    const interval = setInterval(callback, intervalMs);
```
This has been addressed

VRamakrishna commented on examples/cactus-example-cbdc-bridging-frontend/package.json:
```diff
@@ -47,28 +47,56 @@
+    "ts-loader": "9.5.1"
   },
   "engines": {
     "node": ">=18",
```
Here you have the minimum version set to 18. Can you have this constraint be consistent across all the packages?

VRamakrishna commented on packages/cactus-plugin-satp-hermes/src/main/go/generated/gateway-client/git_push.sh:
```diff
@@ -0,0 +1,57 @@
+fi
+
+if [ "$git_user_id" = "" ]; then
+    git_user_id="hyperledger"
```
This should be `hyperledger-cacti`, no?

RafaelAPB replied:
Yes

VRamakrishna commented on packages/cactus-plugin-satp-hermes/src/main/typescript/cross-chain-mechanisms/oracle/smart-contracts/placeholder:
Do you want to call this `.placeholder` as you've done elsewhere? Alternatively, if you want to check in a blank folder, the standard Git convention (as far as I am aware) is to create a blank file named `.gitkeep`.

VRamakrishna commented on packages/cactus-plugin-satp-hermes/src/test/solidity/contracts/satp-contract-interface.sol:
```diff
@@ -0,0 +1,20 @@
+ *  Smart Contract Interface to define the methods needed by SATP Wrapper Contract.
+ */
+
+interface SATPContractInterface {
```
With reference to my earlier comment on this PR, this file seems to be a replica of a file in one of the examples. Ideally, you'd want such an interface to be exported from a common package, so IMO this is the right place for it. And you also ideally don't want to replicate code, which makes maintenance and managing inconsistencies harder.

RafaelAPB replied:
Ideally the examples should import this file, as long as it is a replica. @AndreAugusto11 @LordKubaya

LordKubaya replied:
This interface was removed.

VRamakrishna reviewed (changes requested):
Overall, this looks fine, though I haven't (obviously) gone through each file.

Left some minor questions and suggestions. I'd prefer you address the ones around versioning and naming. Replicated code we can live with as long as eliminating redundancies is in your roadmap and is tracked via issues. Same with moving common code from examples to packages.

General question: I see several instances of generated code (both in JS and Go) being checked in. Are you sure you want to do this? I don't have a strong view on this and will defer to your decision, but just wanted to inquire.

RafaelAPB commented on examples/cactus-example-cbdc-bridging-backend/src/main/typescript/infrastructure/fabric-contracts/satp-contract/chaincode-typescript/package.json:
```diff
@@ -0,0 +1,64 @@
+  "main": "dist/index.js",
+  "typings": "dist/index.d.ts",
+  "engines": {
+    "node": ">=12",
```
We will fix this

RafaelAPB replied:
Yes, this is a typo/copy paste from old fabric samples code.

RafaelAPB commented on packages/cactus-plugin-satp-hermes/src/test/typescript/fabric/contracts/oracle-bl-contract/chaincode-typescript/package.json:
```diff
@@ -0,0 +1,62 @@
+  "main": "dist/index.js",
+  "typings": "dist/index.d.ts",
+  "engines": {
+    "node": ">=12",
```
Agreed

RafaelAPB commented:
Thank you for your feedback, dear @VRamakrishna - we will incorporate those changes and come back to you.

AndreAugusto11 commented on examples/cactus-example-cbdc-bridging-backend/src/main/solidity/satp-contract-interface.sol:
```diff
@@ -0,0 +1,20 @@
+ *  Smart Contract Interface to define the methods needed by SATP Wrapper Contract.
+ */
+
+interface SATPContractInterface {
```
This is a good question. We have just refactored the contracts, and this interface no longer exists. Since we are using a set of JSON files that map protocol actions to token actions, we can avoid these kinds of interfaces that would restrict the tokens that the protocol could use.

RafaelAPB commented:
@VRamakrishna we incorporated your feedback. Could you please re-review?

VRamakrishna reviewed (approved):
Thanks for making the fixes I suggested. LGTM now.

RafaelAPB commented:
The satp-stg branch has a non-linear commit history caused by:

-Duplicate commits with the same messages but different SHAs (from cherry-picks/rebases)
-Parallel fixes developed on side branches that were never merged cleanly.
-Stash commits and force-pushes introducing divergent commit paths.

These changes created a commit graph that cannot be linearized. While we try to [keep history linear](https://stackoverflow.com/questions/20348629/what-are-the-advantages-of-keeping-linear-history-in-git), some work on satp-stg required a combination of rewrites and the events listed above, leading to a non-linear history structure. 

The current structure makes it very difficult to replay commits linearly, even though the branch is up-to-date.
For the sake of time and fixing pressing issues in our codebase, a squash commit will be done, which is the most viable option.


The history of these developments will be preserved:
  - All 138 commits visible in GitHub PR, with the file change history  
  - Complete history backup at: https://github.com/hyperledger-cacti/cacti/tree/satp-stg-10.01.2025
  - Mirror: https://github.com/RafaelAPB/blockchain-integration-framework/tree/satp-stg-10.01.2025
  Either way, the lesson learned is: don't rebase shared branches multiple times!

Lessons learned:
One rebase is fine, multiple are also fine because they are needed from time to time; but each rebase discards old commit IDs and generates new ones. If collaborators or other branches depend on the old commits, Git loses the ability to detect shared history. That’s why it feels like the history is broken, and why squash or merge commits may be needed to reconcile. Therefore, long histories on shared branches are susceptible to the current issue especially when force pushes and cherry picks exist. 

The solution is one that @petermetz told me a long time ago: "small, self-contained PRs".  On shared branches, rebase once at the end when you're confident no more changes needed, and avoid force pushes, that re-write history.

RafaelAPB commented:
Evidence of history divergence

Evidence 1: Duplicate Work on Parallel Branches

Three versions of the same commit exist on different branches:

* a14f89615 ci(satp-hermes): cache yarn.lock  ← Main line
| * b3bb285f4 ci(satp-hermes): cache yarn.lock  ← Parallel branch 1
|/  
| * abcb3e3df ci(satp-hermes): cache yarn.lock  ← Parallel branch 2
|/  
* 33ad00716 fix(satp): fixed client sessiondata for monitoring

Analysis: Same commit message, different SHAs (a14f89615, b3bb285f4, abcb3e3df), 
and DIFFERENT CONTENT - three separate implementations of the same feature:
  - abcb3e3df: Added multi-path yarn.lock caching, changed Docker context to ./
  - b3bb285f4: Fixed Dockerfile path from full path to relative path
  - a14f89615: Combined both changes (final version with both fixes)

These are independent commits created on separate branches that later converged. 
Git cannot determine which version to use during rebase because they contain 
conflicting changes to the same workflow file.

This happens because Git's rebase command replays commits one at a time in linear order. The divergent history creates the following problem:
When replaying a14f89615 (cache yarn.lock), Git doesn't know:
  - Should it include changes from b3bb285f4? (Dockerfile path fix)
  - Should it include changes from abcb3e3df? (multi-path cache + context change)
  - Are these changes already applied or still needed?
  - Which version is "correct" when they modify the same lines differently?

Git sees three different implementations of the same feature with conflicting content. If it picks the wrong one, the CI workflow breaks, so it cannot apply rebase automatically.

━━━━━━━━━

Evidence 2: Multi-Path Docker Fix

Four parallel attempts to fix the same issue:

* 41a142d33 ci(satp-hermes): fix docker image and ci  ← Main line
| * 4f616a45b ci(satp-hermes): fix docker ci and image  ← Branch 1
| * 38c6e534a ci(satp-hermes): fix docker image and ci  ← Branch 2  
| * 0c9ed2f83 ci(satp-hermes): fix docker image and ci  ← Branch 3
|/  
* 420785242 chore(satp-hermes): version 0.0.2-beta release

Analysis: Four different commits (41a142d33, 4f616a45b, 38c6e534a, 0c9ed2f83) with similar 
messages exist on parallel branches. This happened because:
  1. Work was done on the branch and pushed (creating commit 4f616a45b)
  2. Branch was rebased onto main (creating NEW commit 38c6e534a with same changes)
  3. More rebases occurred (creating 0c9ed2f83, then 41a142d33)
  4. Each rebase rewrites history, creating new SHAs for the same logical changes
  5. Old commits remain in the graph even though newer versions exist. Git cannot 
determine which version is "correct" during another rebase operation.

━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

Evidence 3: Documentation Work on Side Branch

* 76aef41b4 feat(satp-hermes): fixed yarn codegen  ← Main development
| * 608adf3f7 docs(satp-hermes): add typedoc documentation  ← Side branch
|/  
* 7492fcbdd fix(satp): fixed error recording for traces

Analysis: Documentation commit 608adf3f7 exists on a side branch that diverged from and 
reconverged with main line, but was never merged (no merge commit exists).

━━━━━━━━━━

Evidence 4: Git Stash Artifacts in History

* 0d59851d9 fix(satp-hermes): update linter config
| *   affedee32 WIP on working-satp-stg: 6fd99e449...  ← Stash commit
| |\  
| | * c02387f85 index on working-satp-stg: 6fd99e449...  ← Stash index
| |/  
| * 6fd99e449 fix(satp-hermes): update linter config  ← Older version
|/  
* 01d25532b chore(security): remove unused dependencies

Analysis: Git stash commits (affedee32, c02387f85) are part of the history, indicating 
work-in-progress was stashed and later reapplied on a different branch. 
Commit 6fd99e449 and 0d59851d9 have the same message but different SHAs.

petermetz commented:
@RafaelAPB Thank you very much for this detailed write-up! I thought I knew all of this already, but reading through it made me realize I have been missing some parts of the full picture until now. :-)
